src/DelegateDecompiler/MethodBodyDecompiler.cs

```diff
@@ -17,35 +17,66 @@ public static LambdaExpression Decompile(MethodInfo method)
         public static LambdaExpression Decompile(MethodInfo method, Type declaringType)
         {
             var args = method.GetParameters()
-                .Select(p => (Address) Expression.Parameter(p.ParameterType, p.Name))
+                .Select(p => (Address)Expression.Parameter(p.ParameterType, p.Name))
                 .ToList();
 
             var methodType = declaringType ?? method.DeclaringType;
             if (!method.IsStatic)
                 args.Insert(0, Expression.Parameter(methodType, "this"));
 
-            var expression = method.IsVirtual
+            var expression = method.IsVirtual && !DecompileExtensions.ConcreteCalls.ContainsKey(new Tuple<object, MethodInfo>(args.FirstOrDefault(), method))
```

Comment:
Here we also check whether a concrete call has been performed for this particular instance